File: sdks/typescript/src/request-control.ts
import { APPLICATION_REQUEST_ID_HEADER } from "./operations";

const MAX_REQUEST_ID_BYTES = 512;
const UTF8_ENCODER = new TextEncoder();

export interface PageOptions {
  size?: number;
  cursor?: string;
}

export interface OperationRequestOptions {
  page?: PageOptions;
  signal?: AbortSignal;
  /** Absolute UTC deadline in microseconds, forwarded to daemon admission. */
  deadlineMicros?: number;
  /** Stable replay identity required by operations whose descriptor says so. */
  requestId?: string;
}

interface RequestControlDescriptor {
  operation: string;
  requestIdControl: "server_minted" | "required";
}

function canonicalRequestId(value: string): boolean {
  return value.length > 0 && value.trim() === value &&
    UTF8_ENCODER.encode(value).byteLength <= MAX_REQUEST_ID_BYTES &&
    !/\p{Cc}/u.test(value);
}

export function applyHttpRequestControls(
  headers: Headers,
  descriptor: RequestControlDescriptor,
  options: OperationRequestOptions,
): string | undefined {
  if (
    options.deadlineMicros !== undefined &&
    (!Number.isSafeInteger(options.deadlineMicros) || options.deadlineMicros <= 0)
  ) return "deadlineMicros must be a positive safe integer";
  if (descriptor.requestIdControl === "required" && options.requestId === undefined) {
    return `${descriptor.operation} requires a stable requestId replay handle`;
  }
  if (descriptor.requestIdControl === "server_minted" && options.requestId !== undefined) {
    return `${descriptor.operation} uses a server-minted request ID`;
  }
  if (options.requestId !== undefined && !canonicalRequestId(options.requestId)) {
    return "requestId is not a canonical request identity";
  }
  if (options.deadlineMicros !== undefined) {
    headers.set("x-tracedecay-deadline-micros", String(options.deadlineMicros));
  }
  if (options.requestId !== undefined) {
    headers.set(APPLICATION_REQUEST_ID_HEADER, options.requestId);
  }
  return undefined;
}

File: crates/tracedecay-sdk/src/request_control.rs
//! Typed HTTP controls shared by generated SDK operations.

use reqwest::header::{HeaderMap, HeaderValue};

use crate::client::ClientError;
use crate::operations::{APPLICATION_REQUEST_ID_HEADER, TypedOperation};

const DEADLINE_HEADER: &str = "x-tracedecay-deadline-micros";

/// Per-invocation transport controls accepted by typed operations.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct OperationRequestOptions {
    /// Absolute UTC deadline in microseconds, forwarded to daemon admission.
    pub deadline_micros: Option<i64>,
    /// Stable caller-owned replay identity required by operations whose
    /// generated request-ID control is `Required`.
    pub request_id: Option<tracedecay_application::RequestId>,
}

pub(crate) fn admit<Operation: TypedOperation>(
    options: &OperationRequestOptions,
) -> Result<Option<tracedecay_application::RequestId>, ClientError> {
    use tracedecay_application::retained_surfaces::SdkRequestIdControlV1;

    match (Operation::REQUEST_ID_CONTROL, options.request_id.as_ref()) {
        (SdkRequestIdControlV1::Required, None) => Err(ClientError::InvalidRequest(format!(
            "{} requires a stable request_id replay handle",
            Operation::OPERATION_ID
        ))),
        (SdkRequestIdControlV1::ServerMinted, Some(_)) => {
            Err(ClientError::InvalidRequest(format!(
                "{} uses a server-minted request ID",
                Operation::OPERATION_ID
            )))
        }
        _ => Ok(options.request_id.clone()),
    }
}

pub(crate) fn apply_http_headers(
    headers: &mut HeaderMap,
    options: OperationRequestOptions,
) -> Result<(), ClientError> {
    if let Some(deadline_micros) = options.deadline_micros {
        if deadline_micros <= 0 {
            return Err(ClientError::InvalidRequest(
                "deadline_micros must be positive".into(),
            ));
        }
        let value = HeaderValue::from_str(&deadline_micros.to_string())
            .map_err(|error| ClientError::InvalidConfiguration(error.to_string()))?;
        headers.insert(DEADLINE_HEADER, value);
    }
    if let Some(request_id) = options.request_id {
        let value = HeaderValue::from_str(request_id.as_str())
            .map_err(|error| ClientError::InvalidConfiguration(error.to_string()))?;
        headers.insert(APPLICATION_REQUEST_ID_HEADER, value);
    }
    Ok(())
}
